<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="tr">
<context>
    <name>QObject</name>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusUtils.cpp" line="118"/>
        <source>Gaps:</source>
        <translation>Boşluklar:</translation>
    </message>
    <message>
        <location filename="../src/misc/FindAlgorithm.cpp" line="160"/>
        <location filename="../src/misc/FindAlgorithm.cpp" line="836"/>
        <source>The search pattern is too long.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>RoughTmCalculatorCmdFactory</name>
    <message>
        <location filename="../src/temperature/rough/RoughTmCalculatorCmdFactory.cpp" line="28"/>
        <source>Rough</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AlignInAminoFormTask</name>
    <message>
        <location filename="../src/util_msaedit/MsaUtilTasks.cpp" line="93"/>
        <source>Align in amino form</source>
        <translation>Amino formunda hizala</translation>
    </message>
    <message>
        <source>Invalid MSA object detected</source>
        <translation type="vanished">Geçersiz MSA nesnesi algılandı</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MsaUtilTasks.cpp" line="103"/>
        <source>AlignInAminoFormTask: Input alphabet is not nucleic!</source>
        <translation>AlignIn Amino Form Görev: Giriş alfabesi nükleik değildir!</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MsaUtilTasks.cpp" line="104"/>
        <source>AlignInAminoFormTask: Input alignment is empty!</source>
        <translation>AlignIn Amino Form Görevi: Giriş hizalaması boş!</translation>
    </message>
    <message>
        <source>Invalid applications settings detected</source>
        <translation type="vanished">Geçersiz uygulama ayarları algılandı</translation>
    </message>
    <message>
        <source>Invalid users applications settings detected</source>
        <translation type="vanished">Geçersiz kullanıcı uygulama ayarları algılandı</translation>
    </message>
    <message>
        <source>Invalid MSA document detected</source>
        <translation type="vanished">Geçersiz MSA belgesi algılandı</translation>
    </message>
    <message>
        <source>NULL clonedObj in AlignInAminoFormTask::prepare!</source>
        <translation type="vanished">Amino Form Görevinde Hizalamada BOŞ klon dObj::hazırlayın!</translation>
    </message>
    <message>
        <source>Can not find row %1 in original alignment.</source>
        <translation type="vanished">Orijinal hizalamada %1 satırı bulunamadı.</translation>
    </message>
</context>
<context>
    <name>U2::AlignmentAlgorithmsRegistry</name>
    <message>
        <location filename="../src/msa_alignment/SimpleAddingToAlignment.cpp" line="222"/>
        <source>Align sequences to alignment with UGENE…</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::AssemblyConsensusAlgorithmFactoryDefault</name>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmDefault.cpp" line="40"/>
        <source>Default</source>
        <translation>Varsayılan</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmDefault.cpp" line="44"/>
        <source>Returns simply the most frequent base and &apos;N&apos; are no reads intersecting this position</source>
        <translation>Basitçe en sık kullanılan tabanı döndürür ve &apos;N&apos; bu konumla kesişen okumalar değildir</translation>
    </message>
</context>
<context>
    <name>U2::AssemblyConsensusAlgorithmFactorySamtools</name>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="45"/>
        <source>SAMtools</source>
        <translation>SAMtools</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="49"/>
        <source>Uses SAMtools to calculate consensus with regard to quality of reads</source>
        <translation>Okuma kalitesiyle ilgili fikir birliğini hesaplamak için SAM araçlarını kullanır</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="70"/>
        <source>Fetching reads from database and converting to SAMtools format</source>
        <translation>Veritabanından okumaları alma ve SAMtools biçimine dönüştürme</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="73"/>
        <source>Sorting reads</source>
        <translation>Okumaları sıralama</translation>
    </message>
    <message>
        <location filename="../src/util_assembly_consensus/AssemblyConsensusAlgorithmSamtools.cpp" line="76"/>
        <source>Calculating consensus</source>
        <translation>Konsensüs hesaplanıyor</translation>
    </message>
</context>
<context>
    <name>U2::BestPositionFindTask</name>
    <message>
        <location filename="../src/msa_alignment/SimpleAddingToAlignment.cpp" line="153"/>
        <source>Best position find task</source>
        <translation>En iyi pozisyon görev bul</translation>
    </message>
</context>
<context>
    <name>U2::CreateSubalignmentTask</name>
    <message>
        <location filename="../src/util_msaedit/CreateSubalignmentTask.cpp" line="58"/>
        <source>Create sub-alignment: %1</source>
        <translation>Alt hizalama oluştur: %1</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/CreateSubalignmentTask.cpp" line="103"/>
        <source>Document is locked: %1</source>
        <translation>Belge kilitli: %1</translation>
    </message>
</context>
<context>
    <name>U2::DnaAssemblyMultiTask</name>
    <message>
        <location filename="../src/misc/DnaAssemblyMultiTask.cpp" line="78"/>
        <source>The short reads can&apos;t be mapped to the reference sequence!</source>
        <translation>Kısa okumalar referans sırasına eşlenemez!</translation>
    </message>
</context>
<context>
    <name>U2::DnaAssemblyToReferenceTask</name>
    <message>
        <location filename="../src/registry/DnaAssemblyTask.cpp" line="37"/>
        <source>Align short reads</source>
        <translation>Kısa okumaları hizalayın</translation>
    </message>
</context>
<context>
    <name>U2::FindAlgorithmTask</name>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="44"/>
        <source>Find in sequence task</source>
        <translation>Sıralı görevde bul</translation>
    </message>
</context>
<context>
    <name>U2::GenomeAssemblyMultiTask</name>
    <message>
        <location filename="../src/misc/GenomeAssemblyMultiTask.cpp" line="80"/>
        <source>Assembly cannot be performed.</source>
        <translation>Montaj yapılamaz.</translation>
    </message>
    <message>
        <location filename="../src/misc/GenomeAssemblyMultiTask.cpp" line="98"/>
        <source>Assembly task finished with error: %1</source>
        <translation>Montaj görevi şu hatayla tamamlandı: %1</translation>
    </message>
    <message>
        <location filename="../src/misc/GenomeAssemblyMultiTask.cpp" line="100"/>
        <source>Assembly task wasn&apos;t set</source>
        <translation>Montaj görevi ayarlanmadı</translation>
    </message>
    <message>
        <location filename="../src/misc/GenomeAssemblyMultiTask.cpp" line="103"/>
        <source>Assembly was finished successfully</source>
        <translation>Montaj başarıyla tamamlandı</translation>
    </message>
    <message>
        <location filename="../src/misc/GenomeAssemblyMultiTask.cpp" line="105"/>
        <source>Assembly failed.</source>
        <translation>Montaj başarısız oldu.</translation>
    </message>
</context>
<context>
    <name>U2::LoadPatternsFileTask</name>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="104"/>
        <source>Load pattern from file</source>
        <translation>Dosyadan desen yükle</translation>
    </message>
    <message>
        <location filename="../src/misc/FindAlgorithmTask.cpp" line="113"/>
        <source>Detecting format error for file %1</source>
        <translation>%1 dosyası için biçim hatası tespit ediliyor</translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryClustal</name>
    <message>
        <source>Emulates ClustalW program and file format behavior.</source>
        <translation type="vanished">ClustalW programını ve dosya biçimi davranışını taklit eder.</translation>
    </message>
    <message>
        <source>ClustalW</source>
        <translation type="vanished">ClustalW</translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryDefault</name>
    <message>
        <source>Based on JalView algorithm. Returns &apos;+&apos; if there are 2 characters with high frequency. Returns symbol in lower case if the symbol content in a row is lower than the threshold specified.</source>
        <translation type="vanished">JalView algoritmasına dayanmaktadır. Yüksek frekanslı 2 karakter varsa &apos;+&apos; döndürür. Bir satırdaki sembol içeriği belirtilen eşikten düşükse sembolü küçük harflerle döndürür.</translation>
    </message>
    <message>
        <source>Default</source>
        <translation type="vanished">Varsayılan</translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryLevitsky</name>
    <message>
        <source>The algorithm proposed by Victor Levitsky to work with DNA alignments.
Collects global alignment frequency for every symbol using extended (15 symbols) DNA alphabet first.
For every column selects the most rare symbol in the whole alignment with percentage in the column greater or equals to the threshold value.</source>
        <translation type="vanished">Victor Levitsky&apos;nin DNA hizalamalarıyla çalışması için önerdiği algoritma.
Önce genişletilmiş (15 sembol) DNA alfabesini kullanarak her sembol için genel hizalama sıklığını toplar.
Her sütun için, sütundaki yüzdesi eşik değerinden daha büyük veya bu değere eşit olacak şekilde, tüm hizalamadaki en nadir sembolü seçer.</translation>
    </message>
    <message>
        <source>Levitsky</source>
        <translation type="vanished">Levitsky</translation>
    </message>
</context>
<context>
    <name>U2::MSAConsensusAlgorithmFactoryStrict</name>
    <message>
        <source>The algorithm returns gap character (&apos;-&apos;) if symbol frequency in a column is lower than threshold specified.</source>
        <translation type="vanished">Bir sütundaki sembol frekansı belirtilen eşikten düşükse algoritma boşluk karakteri (&apos;-&apos;) döndürür.</translation>
    </message>
    <message>
        <source>Strict</source>
        <translation type="vanished">Katı</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithm</name>
    <message>
        <source>MSA distance algorithm &quot;%1&quot; task</source>
        <translation type="vanished">MSA mesafe algoritması &quot;%1&quot; görevi</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactoryHamming</name>
    <message>
        <source>Based on Hamming distance between two sequences</source>
        <translation type="vanished">İki sekans arasındaki Hamming mesafesine göre</translation>
    </message>
    <message>
        <source>Hamming dissimilarity</source>
        <translation type="vanished">Hamming farklılığı</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactoryHammingRevCompl</name>
    <message>
        <source>Based on Hamming distance between two sequences</source>
        <translation type="vanished">İki sekans arasındaki Hamming mesafesine göre</translation>
    </message>
    <message>
        <source>Hamming reverse-complement</source>
        <translation type="vanished">Hamming ters tamamlayıcı</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmFactorySimilarity</name>
    <message>
        <source>Based on similarity distance between two sequences</source>
        <translation type="vanished">İki sekans arasındaki benzerlik mesafesine göre</translation>
    </message>
    <message>
        <source>Similarity</source>
        <translation type="vanished">Benzerlik</translation>
    </message>
</context>
<context>
    <name>U2::MSADistanceAlgorithmHammingRevCompl</name>
    <message>
        <source>An unexpected error has occurred during running the Hamming reverse-complement algorithm.</source>
        <translation type="vanished">Hamming ters tamamlama algoritması çalıştırılırken beklenmeyen bir hata oluştu.</translation>
    </message>
</context>
<context>
    <name>U2::MaConsensusAlgorithmFactorySimpleExtended</name>
    <message>
        <source>The algorithm selects the best character from the extended DNA alphabet. Only bases with frequences which are greater than a threshold value are taken into account.</source>
        <translation type="vanished">Algoritma, genişletilmiş DNA alfabesinden en iyi karakteri seçer. Yalnızca bir eşik değerden daha büyük frekanslara sahip bazlar dikkate alınır.</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MaConsensusAlgorithmSimpleExtended.cpp" line="193"/>
        <source>Simple extended</source>
        <translation>Basit genişletilmiş</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MaConsensusAlgorithmSimpleExtended.cpp" line="194"/>
        <source>The algorithm selects the best character from the extended DNA alphabet. Only bases with frequencies which are greater than a threshold value are taken into account.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::MolecularSurfaceCalcTask</name>
    <message>
        <location filename="../src/molecular_geometry/MolecularSurface.cpp" line="122"/>
        <source>Molecular surface calculation</source>
        <translation>Moleküler yüzey hesabı</translation>
    </message>
</context>
<context>
    <name>U2::MsaColorSchemeRegistry</name>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="566"/>
        <source>No colors</source>
        <translation>Renksiz</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="574"/>
        <source>UGENE Sanger</source>
        <translatorcomment>UGENE Sanger
</translatorcomment>
        <translation>UGENE Sanger</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="578"/>
        <source>Jalview</source>
        <translation>Jalview</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="580"/>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="621"/>
        <source>Percentage identity</source>
        <translation>Yüzde kimliği</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="581"/>
        <source>Percentage identity (colored)</source>
        <translation>Yüzde kimliği (renkli)</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="582"/>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="622"/>
        <source>Percentage identity (gray)</source>
        <translation>Yüzde kimliği (gri)</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="584"/>
        <source>Weak similarities</source>
        <translation>Zayıf benzerlikler</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="595"/>
        <source>Zappo</source>
        <translation>Zappo</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="599"/>
        <source>Tailor</source>
        <translation>Tailor</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="603"/>
        <source>Hydrophobicity</source>
        <translation>Hidrofobiklik</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="607"/>
        <source>Helix propensity</source>
        <translation>Helis eğilimi</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="611"/>
        <source>Strand propensity</source>
        <translation>Şerit eğilimi</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="615"/>
        <source>Turn propensity</source>
        <translation>Eğilimi çevir</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="619"/>
        <source>Buried index</source>
        <translation>Gömülü dizin</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/color_schemes/MsaColorScheme.cpp" line="624"/>
        <source>Clustal X</source>
        <translation>Clustal X</translation>
    </message>
</context>
<context>
    <name>U2::MsaConsensusAlgorithmFactoryClustal</name>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmClustal.cpp" line="32"/>
        <source>ClustalW</source>
        <translation type="unfinished">ClustalW</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmClustal.cpp" line="33"/>
        <source>Emulates ClustalW program and file format behavior.</source>
        <translation type="unfinished">ClustalW programını ve dosya biçimi davranışını taklit eder.</translation>
    </message>
</context>
<context>
    <name>U2::MsaConsensusAlgorithmFactoryDefault</name>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmDefault.cpp" line="33"/>
        <source>Default</source>
        <translation type="unfinished">Varsayılan</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmDefault.cpp" line="34"/>
        <source>Based on JalView algorithm. Returns &apos;+&apos; if there are 2 characters with high frequency. Returns symbol in lower case if the symbol content in a row is lower than the threshold specified.</source>
        <translation type="unfinished">JalView algoritmasına dayanmaktadır. Yüksek frekanslı 2 karakter varsa &apos;+&apos; döndürür. Bir satırdaki sembol içeriği belirtilen eşikten düşükse sembolü küçük harflerle döndürür.</translation>
    </message>
</context>
<context>
    <name>U2::MsaConsensusAlgorithmFactoryLevitsky</name>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmLevitsky.cpp" line="32"/>
        <source>Levitsky</source>
        <translation type="unfinished">Levitsky</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmLevitsky.cpp" line="33"/>
        <source>The algorithm proposed by Victor Levitsky to work with DNA alignments.
Collects global alignment frequency for every symbol using extended (15 symbols) DNA alphabet first.
For every column selects the most rare symbol in the whole alignment with percentage in the column greater or equals to the threshold value.</source>
        <translation type="unfinished">Victor Levitsky&apos;nin DNA hizalamalarıyla çalışması için önerdiği algoritma.
Önce genişletilmiş (15 sembol) DNA alfabesini kullanarak her sembol için genel hizalama sıklığını toplar.
Her sütun için, sütundaki yüzdesi eşik değerinden daha büyük veya bu değere eşit olacak şekilde, tüm hizalamadaki en nadir sembolü seçer.</translation>
    </message>
</context>
<context>
    <name>U2::MsaConsensusAlgorithmFactoryStrict</name>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmStrict.cpp" line="35"/>
        <source>Strict</source>
        <translation type="unfinished">Katı</translation>
    </message>
    <message>
        <location filename="../src/util_msa_consensus/MsaConsensusAlgorithmStrict.cpp" line="36"/>
        <source>The algorithm returns gap character (&apos;-&apos;) if symbol frequency in a column is lower than threshold specified.</source>
        <translation type="unfinished">Bir sütundaki sembol frekansı belirtilen eşikten düşükse algoritma boşluk karakteri (&apos;-&apos;) döndürür.</translation>
    </message>
</context>
<context>
    <name>U2::MsaDistanceAlgorithm</name>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithm.cpp" line="59"/>
        <source>MSA distance algorithm &quot;%1&quot; task</source>
        <translation type="unfinished">MSA mesafe algoritması &quot;%1&quot; görevi</translation>
    </message>
</context>
<context>
    <name>U2::MsaDistanceAlgorithmFactoryHamming</name>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmHamming.cpp" line="33"/>
        <source>Based on Hamming distance between two sequences</source>
        <translation type="unfinished">İki sekans arasındaki Hamming mesafesine göre</translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmHamming.cpp" line="37"/>
        <source>Hamming dissimilarity</source>
        <translation type="unfinished">Hamming farklılığı</translation>
    </message>
</context>
<context>
    <name>U2::MsaDistanceAlgorithmFactoryHammingRevCompl</name>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmHammingRevCompl.cpp" line="38"/>
        <source>Based on Hamming distance between two sequences</source>
        <translation type="unfinished">İki sekans arasındaki Hamming mesafesine göre</translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmHammingRevCompl.cpp" line="42"/>
        <source>Hamming reverse-complement</source>
        <translation type="unfinished">Hamming ters tamamlayıcı</translation>
    </message>
</context>
<context>
    <name>U2::MsaDistanceAlgorithmFactorySimilarity</name>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmSimilarity.cpp" line="33"/>
        <source>Based on similarity distance between two sequences</source>
        <translation type="unfinished">İki sekans arasındaki benzerlik mesafesine göre</translation>
    </message>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmSimilarity.cpp" line="37"/>
        <source>Similarity</source>
        <translation type="unfinished">Benzerlik</translation>
    </message>
</context>
<context>
    <name>U2::MsaDistanceAlgorithmHammingRevCompl</name>
    <message>
        <location filename="../src/util_msa_distance/MsaDistanceAlgorithmHammingRevCompl.cpp" line="72"/>
        <source>An unexpected error has occurred during running the Hamming reverse-complement algorithm.</source>
        <translation type="unfinished">Hamming ters tamamlama algoritması çalıştırılırken beklenmeyen bir hata oluştu.</translation>
    </message>
</context>
<context>
    <name>U2::MsaHighlightingSchemeRegistry</name>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="114"/>
        <source>No highlighting</source>
        <translation>Vurgulama yok</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="115"/>
        <source>Agreements</source>
        <translation>Anlaşmalar</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="116"/>
        <source>Disagreements</source>
        <translation>Anlaşmazlıklar</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="117"/>
        <source>Gaps</source>
        <translation>Boşluklar</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="118"/>
        <source>Conservation level</source>
        <translation>Koruma seviyesi</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="119"/>
        <source>Transitions</source>
        <translation>Geçişler</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/highlighting_schemes/MsaHighlightingScheme.cpp" line="120"/>
        <source>Transversions</source>
        <translation>Dönüşümler</translation>
    </message>
</context>
<context>
    <name>U2::ORFFindTask</name>
    <message>
        <location filename="../src/util_orf/ORFAlgorithmTask.cpp" line="31"/>
        <source>ORF find</source>
        <translation>ORF bul</translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryBVH</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmBVH.cpp" line="40"/>
        <source>Berg and von Hippel weight function</source>
        <translation>Berg ve von Hippel ağırlık fonksiyonu</translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryLOD</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmLOD.cpp" line="37"/>
        <source>log-odds</source>
        <translation>günlük oranlar</translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmLOD.cpp" line="41"/>
        <source>log-odds weight function</source>
        <translation>günlük oran ağırlık işlevi</translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryMCH</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmMCH.cpp" line="36"/>
        <source>Match</source>
        <translation>Eşleştir</translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmMCH.cpp" line="40"/>
        <source>Match weight function</source>
        <translation>Ağırlık işlevi eşleştir</translation>
    </message>
</context>
<context>
    <name>U2::PWMConversionAlgorithmFactoryNLG</name>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmNLG.cpp" line="36"/>
        <source>NLG</source>
        <translation>NLG</translation>
    </message>
    <message>
        <location filename="../src/util_weight_matrix/PWMConversionAlgorithmNLG.cpp" line="40"/>
        <source>NLG weight function</source>
        <translation>NLG ağırlık işlevi</translation>
    </message>
</context>
<context>
    <name>U2::PairwiseAlignmentTask</name>
    <message>
        <location filename="../src/pairwise_alignment/PairwiseAlignmentTask.cpp" line="45"/>
        <source>Pairwise alignment task</source>
        <translation>İkili hizalama görevi</translation>
    </message>
</context>
<context>
    <name>U2::PhyTreeGeneratorLauncherTask</name>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="46"/>
        <source>Calculating Phylogenetic Tree</source>
        <translation>Filogenetik Ağacın Hesaplanması</translation>
    </message>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="56"/>
        <source>Tree algorithm %1 is not found</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="81"/>
        <source>Failed to map row name: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Tree construction algorithm %1 not found</source>
        <translation type="vanished">Ağaç oluşturma algoritması %1 bulunamadı</translation>
    </message>
</context>
<context>
    <name>U2::PhyTreeGeneratorTask</name>
    <message>
        <location filename="../src/phyltree/PhyTreeGeneratorTask.cpp" line="33"/>
        <source>Calculating Phylogenetic Tree</source>
        <translation>Filogenetik Ağacın Hesaplanması</translation>
    </message>
</context>
<context>
    <name>U2::SecStructPredictTask</name>
    <message>
        <location filename="../src/registry/SecStructPredictTask.cpp" line="27"/>
        <source>Secondary structure predict</source>
        <translation>İkincil yapı tahmini</translation>
    </message>
</context>
<context>
    <name>U2::SequenceContentFilterTask</name>
    <message>
        <location filename="../src/misc/SequenceContentFilterTask.cpp" line="39"/>
        <source>Sequence content</source>
        <translation>Sıra içeriği</translation>
    </message>
</context>
<context>
    <name>U2::SimpleAddToAlignmentTask</name>
    <message>
        <source>Align sequences to an existing alignment by UGENE started</source>
        <translation type="vanished">UGENE tarafından başlatılan dizileri mevcut bir hizalamaya hizalayın</translation>
    </message>
    <message>
        <location filename="../src/msa_alignment/SimpleAddingToAlignment.cpp" line="51"/>
        <source>Align sequences to alignment with UGENE started</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <location filename="../src/msa_alignment/SimpleAddingToAlignment.cpp" line="136"/>
        <source>The following sequence(s) were not aligned as they do not contain meaningful characters: &quot;%1&quot;.</source>
        <translation>Aşağıdaki diziler, anlamlı karakterler içermedikleri için hizalanmamıştır: &quot;%1&quot;.</translation>
    </message>
</context>
<context>
    <name>U2::SmithWatermanReportCallbackAnnotImpl</name>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="64"/>
        <source>Annotation object not found.</source>
        <translation>Ek açıklama nesnesi bulunamadı.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="68"/>
        <source>Annotation table is read-only</source>
        <translation>Ek açıklama tablosu salt okunurdur</translation>
    </message>
</context>
<context>
    <name>U2::SmithWatermanReportCallbackMAImpl</name>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="208"/>
        <source>SmithWatermanReportCallback failed to create new MA document</source>
        <translation>SmithWatermanReportCallback, yeni MA belgesi oluşturamadı</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="226"/>
        <source>Failed to add a reference subsequence row.</source>
        <translation>Referans alt sıra satırı eklenemedi.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="230"/>
        <source>Failed to add a pattern subsequence row.</source>
        <translation>Kalıp alt dizisi satırı eklenemedi.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="233"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="318"/>
        <source>Failed to create an alignment.</source>
        <translation>Hizalama oluşturulamadı.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="261"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="335"/>
        <source>Smith-Waterman pairwise alignment: no results</source>
        <translation>Smith-Waterman ikili hizalama: sonuç yok</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="268"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="340"/>
        <source>Sequences aren&apos;t stored in the source msa dbi</source>
        <translation>Diziler kaynak msa dbi&apos;de depolanmaz</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="284"/>
        <source>SmithWatermanReportCallback failed to create new MA document.</source>
        <translation>SmithWatermanReportCallback, yeni MA belgesi oluşturamadı.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="290"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="292"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="349"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="351"/>
        <source>Failed to get the sequence object.</source>
        <translation>Sıra nesnesi alınamadı.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="306"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="308"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="365"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="369"/>
        <source>Failed to get the sequence data.</source>
        <translation>Sıra verileri alınamadı.</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="378"/>
        <source>Failed to get msa from dbi</source>
        <translation>Dbi&apos;den msa alınamadı</translation>
    </message>
    <message>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="401"/>
        <location filename="../src/smith_waterman/SmithWatermanReportCallback.cpp" line="403"/>
        <source>Failed to update row gap model</source>
        <translation>Satır aralığı modeli güncellenemedi</translation>
    </message>
</context>
<context>
    <name>U2::SplicedAlignmentTask</name>
    <message>
        <source>SplicedAlignmentTask</source>
        <translation type="vanished">Eklenmiş Hizalama Görevi</translation>
    </message>
    <message>
        <location filename="../src/registry/SplicedAlignmentTask.cpp" line="27"/>
        <source>Partial Alignment Task</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::SubstMatrixRegistry</name>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="102"/>
        <source>Reading substitution matrix from %1</source>
        <translation>%1&apos;den ikame matrisi okunuyor</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="110"/>
        <source>Error weight matrix file &apos;%1&apos; : %2</source>
        <translation>Hata ağırlık matrisi dosyası &apos;%1&apos;: %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="118"/>
        <source>Error opening file for read: %1</source>
        <translation>Dosyayı okumak için açma hatası: %1</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="125"/>
        <source>Error reading file: %1</source>
        <translation>Dosya okunurken hata oluştu: %1</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="158"/>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="182"/>
        <source>Invalid character token &apos;%1&apos; , line %2</source>
        <translation>Geçersiz karakter simgesi &apos;%1&apos; , satır %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="163"/>
        <source>Duplicate character &apos;%1&apos; , line %2</source>
        <translation>Yinelenen karakter &apos;%1&apos;, satır %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="170"/>
        <source>Alphabet is neither nucleic nor protein!</source>
        <translation>Alfabe ne çekirdek ne de protein!</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="174"/>
        <source>Matrix doesn&apos;t contains score for default alphabet character &apos;%1&apos;</source>
        <translation>Matris, varsayılan alfabe karakteri &apos;%1&apos; için puan içermiyor</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="187"/>
        <source>Invalid character row &apos;%1&apos; , line %2</source>
        <translation>Geçersiz karakter satırı &apos;%1&apos;, satır %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="191"/>
        <source>Duplicate character mapping &apos;%1&apos; , line %2</source>
        <translation>Yinelenen karakter eşlemesi &apos;%1&apos;, satır %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="196"/>
        <source>Invalid number of columns &apos;%1&apos; , line %2</source>
        <translation>Geçersiz sütun sayısı &apos;%1&apos;, satır %2</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="204"/>
        <source>Can&apos;t parse numeric value &apos;%1&apos;, line %2</source>
        <translation>Sayısal değer &apos;%1&apos;, satır %2 ayrıştırılamıyor</translation>
    </message>
    <message>
        <location filename="../src/registry/SubstMatrixRegistry.cpp" line="213"/>
        <source>Unexpected end of file!</source>
        <translation>Beklenmedik dosya sonu!</translation>
    </message>
</context>
<context>
    <name>U2::TranslateMsa2AminoTask</name>
    <message>
        <location filename="../src/util_msaedit/MsaUtilTasks.cpp" line="46"/>
        <location filename="../src/util_msaedit/MsaUtilTasks.cpp" line="58"/>
        <source>Translate nucleic alignment to amino</source>
        <translation>Nükleik hizalamayı aminoya çevirin</translation>
    </message>
    <message>
        <source>Invalid MSA object detected</source>
        <translation type="vanished">Geçersiz MSA nesnesi algılandı</translation>
    </message>
    <message>
        <source>Multiple alignment already has amino-acid alphabet</source>
        <translation type="vanished">Çoklu hizalamada zaten amino asit alfabesi var</translation>
    </message>
    <message>
        <location filename="../src/util_msaedit/MsaUtilTasks.cpp" line="52"/>
        <source>Unable to find suitable translation for %1</source>
        <translation>%1 için uygun çeviri bulunamadı</translation>
    </message>
    <message>
        <source>Invalid translation object</source>
        <translation type="vanished">Geçersiz çeviri nesnesi</translation>
    </message>
</context>
</TS>
